src/core/LocklessAllocator.cpp

```diff
@@ -101,27 +101,27 @@ static int ffs( int i )
 
 void * LocklessAllocator::alloc()
 {
-	int available;
+	// Some of these CAS loops could probably use relaxed atomics, as discussed
+	// in http://en.cppreference.com/w/cpp/atomic/atomic/compare_exchange.
+	// Let's use sequentially-consistent ops to be safe for now.
+	int available = m_available.load();
 	do
 	{
-		available = m_available;
 		if( !available )
 		{
 			fprintf( stderr, "LocklessAllocator: No free space\n" );
 			return NULL;
 		}
 	}
-	while( !m_available.testAndSetOrdered( available, available - 1 ) );
+	while (!m_available.compare_exchange_weak(available, available - 1));
 
-	size_t startIndex = m_startIndex.fetchAndAddOrdered( 1 )
-							% m_freeStateSets;
-	for( size_t set = startIndex;; set = ( set + 1 ) % m_freeStateSets )
+	const size_t startIndex = ++m_startIndex % m_freeStateSets;
```

Comment:
Shouldn't the above be post-increment: `m_startIndex++`?

Reply:
Yes.